enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,754 @@
+It is expected that this would not cause issue in most cases as the infrastructure provider should be able to handle multiple requests to delete the same object.
+Where Machine lifecycle hooks are in place on Machines, there should be mirrored across both APIs and should prevent either taking action until they are removed.
+Machine lifecycle hooks are expected to continue to function as expected.
+
+From this, it is imperative that we ensure Create operations are not reconciled by multiple Machine controllers to prevent the potential of leaking resources.
+Testing will be focused on the creation flow.
+
+[filtering]: #filtering-reconciles-in-capi-controllers
+
+### Drawbacks
+
+#### Carrying filtering logic in the Cluster API controllers
+
+We aim to carry as little code in forks as possible, however, we are changing the behaviour of the controllers with the reconcile filtering logic described above.
+
+This will be a fairly major patch that we need to carry.
+We need to make sure that the patch will apply cleanly in the future and that we minimise the maintenance burden of this carry as much as possible.
+
+#### Mirroring resources may be confusing for end users
+
+We want to provide an easy method for migration for users, however, having multiple representations in OpenShift of the same infrastructure, may be confusing for end users.
+
+They may expect to be able to make changes to either resource and have them reflected.
+For that to be safe we would have to make sure all writes to objects are synchronous to both objects.
+This is technically very challenging and has been dismissed as an alternative.
+
+Documentation will be provided to explain to users the expected behaviours and the expectations to allow them to migrate their infrastructure across.
+
+## Design Details
+
+### Open Questions
+
+Q. How complex will it be to have the Cluster API controllers look up Machine API resources and prevent reconciliation when not authoritative? What will the maintenance burden be for this?
+A. TODO
+
+### Test Plan
+
+#### Conversion of provider specs
+
+The conversion of provider specs from Machine API to Cluster API should be able to be statically tested. We will build out extensive conversion testing via unit tests to ensure that, with a desired input, the correct conversion happens for the provider specs and the associated resources.
+
+#### Authoritative API reconciliation
+
+Machine API and Cluster API controllers already use envtest as a way to provide integration testing with a "real" API server.
+We can extend the tests here to account for the cases where the authoritative API annotation feature is in use.
+
+By extending the integration tests, we can sculpt the desired inputs for mirror resources existing, not existing, and authoritative in both API versions,
+and then determine whether the reconcile continued or exited based on these conditions.
+
+#### Full migration E2E
```

Comment:
I wanted to write here:

We can set a cluster-wide flag to make CAPI authoritative in all cases then run our existing test suites, which should continue to pass.

However, we can't do that because our existing test suites will continue to use MAPI, and we don't allow continued synchronisation of resources from MAPI->CAPI when CAPI is authoritative.

This suggests a couple of problems to me:
* We can't leverage our existing battle-hardened test suite as a regression suite, at least for machine operations
* We can't make CAPI globally authoritative until all our users stop using MAPI, which we may never achieve

Reply:
> We can't leverage our existing battle-hardened test suite as a regression suite, at least for machine operations

Given the existing tests that exist in origin, we can update these to use unstructured since they only leverage MachineSet scaling.

For other test suites, you're correct.

However, one of the plans for the future is that we eventually turn the feature into a one way sync. WIth this (feature gated) approach we can still run the existing test suite, leveraging the MAPI suite with everything being run by CAPI controllers

This is also how we will make CAPI the backend globally eventually, but we have no intention of forcing CAPI authoritative in the sense of the API at least, as a part of this project